### ISSUE TITLE
Format Converter + ECC Content Scripts

### ISSUE THREAD

**Issue by JPerreault01** [JPerreault01/Coffee-Bean-Scraper#50]
@claude

# Claude Code Task: Format Converter + ECC Content Scripts

This task has two parts. Complete both before committing.

---

## Part 1: Build format_for_finetuning.py

Build the file `data_pipeline/format_for_finetuning.py`.

### What this script does
Converts cleaned JSONL data from `training_data/cleaned/` into ChatML instruction
pairs ready for fine-tuning a Hermes-3 or Llama-3.1-8B model. This is Phase 1 of
the fine-tuning pipeline.

For each cleaned record that passes quality filters, it calls the Anthropic API
(claude-haiku-4-5-20251001) to generate the user-side prompt that would have produced
the cleaned text as a response. The resulting (system, user, assistant) triple is the
instruction pair.

The script processes all three sources: reddit, web, and youtube.

### Input
```
training_data/cleaned/
├── reddit/        ← one .jsonl per subreddit
├── web/           ← one .jsonl per site
└── youtube/       ← one .jsonl per channel
```

### Output
```
training_data/finetune/
├── train.jsonl         ← 90% of pairs, ChatML format
├── val.jsonl           ← 10% of pairs, ChatML format
├── stats.json          ← counts by source, category, split
└── checkpoint.json     ← processed source_ids for resume support
```

### ChatML output format

Each line in train.jsonl and val.jsonl is a JSON object:
```json
{
  "messages": [
    {"role": "system", "content": "..."},
    {"role": "user", "content": "..."},
    {"role": "assistant", "content": "..."}
  ],
  "meta": {
    "source": "reddit|web|youtube",
    "site_or_channel": "barista_hustle|espresso|james_hoffmann|...",
    "category": "espresso|grinders|pour_over|...",
    "quality_score": 0.87,
    "source_id": "unique identifier from the raw record"
  }
}
```

`site_or_channel` is derived from the JSONL filename stem (e.g. `barista_hustle.jsonl`
→ `"barista_hustle"`, `james_hoffmann.jsonl` → `"james_hoffmann"`,
`espresso.jsonl` → `"espresso"`).

### System prompt (use this exactly for all pairs)

```
You are a coffee expert writing for a niche review and knowledge site. You have deep
technical knowledge of espresso, filter brewing, grinders, origins, and roast science.

Your voice is direct, confident, and specific. No hedging. No filler. No marketing
language. Short declarative sentences. Specific claims over vague ones — "turns acrid
past 205°F" not "can be harsh". You apply critical judgment: clean finishes over
lingering bitterness, forgiving brew profiles over finicky ones, value over brand
premiums.

When reviewing products, the coffee is the subject. State what it is and what it does.
Second person ("you get", "you'll find") is preferred. Never claim to have personally
tried a product unless explicitly told this is a personal review.
```

### Quality filters — apply to ALL sources before processing

Pass a record only if ALL of the following are true:
- `quality_score >= 0.65` (configurable via --min-quality flag)
- `flagged` is absent or false
- `duplicate` is absent or false
- `is_english` is absent or true
- Primary text length meets the source-specific minimum (see per-source rules below)

#### Reddit
- Primary text: `raw.title` + " " + `raw.body`
- Minimum text length: 150 chars
- Skip records where `raw.title` matches any of these spam patterns (case-insensitive):
  starts with "price check", "[wts]", "[wtb]", "where to buy", "iso "
- Truncate assistant response to 1500 chars if longer
- source_id: use `raw.post_id` if present, otherwise SHA-256 hex digest of title+body
  truncated to 16 chars

#### Web
- Primary text: `raw.body`
- Minimum text length: 400 chars
- Use full body as assistant response, truncated to 2500 chars
- source_id: use `raw.url` if present, otherwise `raw.site` + "_" + str(index)

#### YouTube
- Primary text: `raw.transcript`
- Minimum text length: 300 chars
- Chunk transcripts: split into overlapping chunks of 3200 chars with 200 char overlap.
  Each chunk becomes one pair. Use first 2500 chars of each chunk as the assistant
  response.
- source_id: `raw.video_id` + "_" + str(chunk_index) (e.g. "dQw4w9WgXcQ_0")

### Category assignment

Assign the primary category from `domain_tags` using this priority order (first match
wins):
1. espresso
2. grinders
3. pour_over
4. french_press
5. cold_brew
6. roast
7. origins
8. equipment_review
9. troubleshooting
10. beginner

If none match, category = "general". Include general-tagged records in the output.

### Backtranslation API call

Instantiate the Anthropic client explicitly:
```python
client = anthropic.Anthropic(api_key=os.environ["CLAUDE_API_KEY"])
```

For each passing record, call the API with these parameters:
- Model: `claude-haiku-4-5-20251001`
- Max tokens: 150
- Temperature: 0.7

Prompt:
```
You are generating training data for a coffee knowledge model.

Below is a piece of coffee content. Write a single natural question or instruction
that someone might ask, where this content would be an ideal response. The question
should be specific to the content — not generic. Match the technical level of the
content. Do not ask about specific author names, dates, or publication sources.
Return only the question, no preamble.

Content category: {category}
Content source type: {source_type} (reddit discussion / web article / youtube transcript)

Content:
{assistant_text}
```

### Checkpoint and resume

On each successful pair generation, append the `source_id` to
`training_data/finetune/checkpoint.json` (a JSON list, written atomically by reading,
appending, and rewriting the file).

On startup, load checkpoint.json and skip any record whose source_id is already in it.

`--fresh` clears the entire checkpoint file on startup regardless of which `--source`
is specified. If the user wants to reprocess only one source without clearing others,
they must edit checkpoint.json manually.

### Train/val split

After all pairs are generated:
- Shuffle with fixed random seed (seed=42)
- Split 90% train / 10% val, stratified by category
- If a category has fewer than 5 records, put all of them in train
- Write train.jsonl, val.jsonl, stats.json

stats.json structure:
```json
{
  "total_pairs": 0,
  "train_pairs": 0,
  "val_pairs": 0,
  "by_source": {"reddit": 0, "web": 0, "youtube": 0},
  "by_category": {"espresso": 0, "...": 0},
  "by_split": {"train": {"espresso": 0, "...": 0}, "val": {"espresso": 0, "...": 0}}
}
```

### Rate limiting and retry

- Default delay between API calls: 0.3 seconds. Configurable via `--delay` flag.
- On HTTP 429 (rate limit): wait 10 seconds and retry up to 3 times before skipping
  the record and logging a warning.
- On any other API error after 3 retries: log a warning with source_id, skip, continue.

### Progress display

Print to stdout (not logging):
```
[reddit] espresso.jsonl: 42/180 | passed: 38 | filtered: 4
[web] barista_hustle.jsonl: 12/45 | passed: 12 | filtered: 0
Pairs generated: 187 | Estimated cost: $0.019
```

Estimated cost: (total_api_calls * 750 * 0.80 / 1_000_000) input +
(total_api_calls * 150 * 4.00 / 1_000_000) output, using Haiku pricing at
$0.80/M input and $4.00/M output tokens. Update cost line in place using `\r`.

### CLI
```
python data_pipeline/format_for_finetuning.py
python data_pipeline/format_for_finetuning.py --source reddit
python data_pipeline/format_for_finetuning.py --source web
python data_pipeline/format_for_finetuning.py --source youtube
python data_pipeline/format_for_finetuning.py --min-quality 0.70
python data_pipeline/format_for_finetuning.py --delay 0.5
python data_pipeline/format_for_finetuning.py --fresh
python data_pipeline/format_for_finetuning.py --dry-run
```

`--dry-run`: apply all filters, count pairs that would be generated including YouTube
chunks, print breakdown by source and estimated cost, make zero API calls, write no
output files.

### Environment
Read API key from `CLAUDE_API_KEY`. If not set, print a clear error message and exit
with code 1.

### Error handling
- `training_data/cleaned/` does not exist: exit with clear message
- Create `training_data/finetune/` if it does not exist
- Malformed record: log warning, skip, continue
- API failure after retries: log warning with source_id, skip, continue

### Code style
- Single `main()` entry point
- Functions for each major section
- Type hints throughout
- Docstrings on all functions
- Python 3.10+ compatible
- `logging` for diagnostics; `print` only for progress counter and final summary

### Dependencies
`anthropic>=0.25.0` is already in requirements.txt. Standard library for everything
else.

---

## Part 2: Build three ECC-powered content scripts

Before writing any code for this part, fetch and read the following files from the
ECC repository. They define the behavioral patterns to embed in each script:

```
https://raw.githubusercontent.com/affaan-m/ECC/main/skills/article-writing/SKILL.md
https://raw.githubusercontent.com/affaan-m/ECC/main/skills/content-engine/SKILL.md
https://raw.githubusercontent.com/affaan-m/ECC/main/skills/market-research/SKILL.md
```

Use `curl` or `requests` to fetch them. Read them before writing the scripts. The
skill files define the behavioral instructions — extract the core workflow and
prompting patterns from each and embed them as system prompts in the API calls below.
Do not summarize or paraphrase the skill instructions; use them directly.

Build these three files:

---

### 2a. `scrapers/write_review.py`

Uses the `article-writing` skill to generate a full bean review in the site's voice.

#### Inputs
- Product ID from `scrapers/products.json` (passed as `--product <id>`)
- The site's writing standards, read from `CLAUDE.md` in the repo root at runtime
- Optional: RAG context snippets passed via `--context-file <path>` (plain text,
  injected into the prompt)

#### Behavior
- Load the product record from products.json by ID
- Read CLAUDE.md to extract the review format and writing style sections
- Read the article-writing SKILL.md behavioral instructions (embed as system prompt)
- Call `claude-sonnet-4-5-20251001` (not Haiku — reviews need quality) with:
  - System: the article-writing skill instructions + the site's voice and format rules
    extracted from CLAUDE.md
  - User: a prompt that provides the product spec and requests a complete review
    in the established format
- Write the draft to `drafts/<product-id>-review.md`
- Print the output path on success

#### Review format to enforce (embed this in the system prompt verbatim)
```
## [Product Name] Review

**One-line verdict**: [Direct, specific, no hedge words]

| Spec | Detail |
|---|---|
| Roast | |
| Origin | |
| Process | |
| Best for | [brew methods] |
| Price/oz | $X.XX |

### Tasting notes
- [Specific note]
- [3–5 bullets total]

### Who it's for
[1–2 sentences. Specific.]

### Who should skip it
[1–2 sentences. Honest.]

### Price analysis
[Current price vs 30-day average, value judgment]

### Rating: X/10
[One sentence explaining the score]
```

#### Writing rules to enforce (embed verbatim)
- Direct, slightly opinionated voice
- No filler phrases: "in conclusion", "it's worth noting", "at the end of the day"
- No fake hedging: "some may find", "could potentially"
- Specific over vague: details beat generalities
- Short sentences preferred

#### CLI
```
python scrapers/write_review.py --product lavazza-super-crema
python scrapers/write_review.py --product lavazza-super-crema --context-file path/to/context.txt
python scrapers/write_review.py --product lavazza-super-crema --mock
```

`--mock` prints a template with the product fields filled in but no generated text.
Read API key from `CLAUDE_API_KEY`.

---

### 2b. `scrapers/repurpose.py`

Uses the `content-engine` skill to repurpose a finished review into derivative formats.

#### Inputs
- Path to a finished review draft (passed as `--review <path>`)
- `--formats` flag accepting one or more of: `email`, `social`, `all` (default: `all`)

#### Behavior
- Read the review file
- Read the content-engine SKILL.md behavioral instructions (embed as system prompt)
- Make one API call per requested format using `claude-haiku-4-5-20251001`:

  **Email format** (for Beehiiv price-drop alerts):
  - 3–5 sentence teaser
  - Ends with a clear CTA and placeholder affiliate link `[LINK]`
  - Subject line included
  - Tone: informative, not promotional

  **Social format** (Twitter/X):
  - Under 280 chars
  - Lead with the one-line verdict
  - Include price if available from the review
  - No hashtags unless they appear naturally

- Write outputs to `drafts/<review-stem>-email.md` and/or
  `drafts/<review-stem>-social.txt`
- Print output paths on success

#### CLI
```
python scrapers/repurpose.py --review drafts/lavazza-super-crema-review.md
python scrapers/repurpose.py --review drafts/lavazza-super-crema-review.md --formats email
python scrapers/repurpose.py --review drafts/lavazza-super-crema-review.md --formats social
python scrapers/repurpose.py --review drafts/lavazza-super-crema-review.md --formats all
```

Read API key from `CLAUDE_API_KEY`.

---

### 2c. `scrapers/market_research.py`

Uses the `market-research` skill to build a source-attributed price analysis section
for a product.

#### Inputs
- Product ID from `scrapers/products.json` (passed as `--product <id>`)
- Optional: `--days <n>` for how many days of price history to consider (default: 30)

#### Behavior
- Load the product record from products.json by ID
- Query the local SQLite database at `data/prices.db` for price history:
  `SELECT price, checked_at FROM price_history WHERE product_id = ? ORDER BY checked_at DESC LIMIT 200`
- Compute: current price, 30-day average, 30-day low, 30-day high, price trend
  (rising/falling/flat based on last 7 days vs prior 7 days)
- Read the market-research SKILL.md behavioral instructions (embed as system prompt)
- Call `claude-haiku-4-5-20251001` with the price data to generate a 2–3 paragraph
  price analysis section that:
  - States the current price and how it compares to the 30-day average
  - Notes the best price seen in the window and whether now is a good time to buy
  - References the Amazon or roaster URL from products.json
  - Attributes all claims to "price history data" (no invented sources)
- Write output to `drafts/<product-id>-price-analysis.md`
- If `data/prices.db` does not exist or has no rows for the product, generate a
  placeholder section noting that price tracking is not yet active for this product

#### CLI
```
python scrapers/market_research.py --product lavazza-super-crema
python scrapers/market_research.py --product lavazza-super-crema --days 60
```

Read API key from `CLAUDE_API_KEY`.

---

## General requirements for all Part 2 scripts

- Read API key from `CLAUDE_API_KEY`; exit with code 1 and a clear message if absent
- Create `drafts/` directory if it does not exist
- `logging` for diagnostics, `print` for final output paths
- Type hints and docstrings throughout
- Python 3.10+ compatible
- Standard library only except `anthropic` (already in requirements.txt) and
  `sqlite3` (stdlib)
- Do not hardcode any product IDs, file paths, or API keys

---

## Commit message

```
feat: add format_for_finetuning.py and ECC-powered content scripts (write_review, repurpose, market_research)
```
